### PR TITLE
Migrate typeck::outlives::mod.rs to SessionDiagnostic

### PR DESCRIPTION
**Pull request by ameknite**
Migration of a part of the outlive module to Session Diagnostic. 
This is my first PR, I hope is ok

r? rust-lang/diagnostics

### CODE REVIEW

File: compiler/rustc_typeck/src/errors.rs [hunk 1]
```diff
@@ -324,3 +324,20 @@ pub struct SubstsOnOverriddenImpl {
     #[primary_span]
     pub span: Span,
 }
+
+pub struct RustcOutlives {
+    pub span: Span,
+    pub pred: Vec<String>,
+}
+
+impl SessionDiagnostic<'_> for RustcOutlives {
```
Comment (compiler-errors) [inline compiler/rustc_typeck/src/errors.rs:333]: This should not use a manual implementation of `SessionDiagnostic` but instead be using the derive macro, I think.

Comment (davidtwco) [reply compiler/rustc_typeck/src/errors.rs:333]: Yeah, you will be able to simplify this with a diagnostic derive, you can see examples [in the blog post](https://blog.rust-lang.org/inside-rust/2022/08/16/diagnostic-effort.html#using-a-diagnostic-derive).

File: compiler/rustc_typeck/src/outlives/mod.rs [hunk 2]
```diff
@@ -59,12 +61,7 @@ fn inferred_outlives_of(tcx: TyCtxt<'_>, item_def_id: DefId) -> &[(ty::Predicate
                         .collect();
                     pred.sort();
 
-                    let span = tcx.def_span(item_def_id);
-                    let mut err = tcx.sess.struct_span_err(span, "rustc_outlives");
```
Comment (davidtwco) [inline compiler/rustc_typeck/src/outlives/mod.rs:63]: pre-existing: but this is a bizarre error message.. and the blame is from many years ago.

**Pull request closed without merging**